### ISSUE TITLE
Dataclass schema regression in 1.9.0

### ISSUE THREAD

**Issue by pgjones** [samuelcolvin/pydantic#3658]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.9.0
            pydantic compiled: True
                 install path: venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.0 (default, Oct  9 2021, 15:30:15) [Clang 13.0.0 (clang-1300.0.29.3)]
                     platform: macOS-11.6-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```
The following errors, (with Pydantic 1.9.0)
```python
from pydantic.dataclasses import dataclass as pydantic_dataclass
from pydantic.schema import model_schema 
from typing import List
from dataclasses import dataclass

@dataclass
class Rule:
    name: str

@dataclass
class RuleData:
    rule: List[Rule]

model_schema(pydantic_dataclass(RuleData))
```
The error is,
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/schema.py", line 185, in pydantic.schema.model_schema
  File "pydantic/schema.py", line 617, in pydantic.schema.model_process_schema
  File "pydantic/schema.py", line 658, in pydantic.schema.model_type_schema
  File "pydantic/schema.py", line 258, in pydantic.schema.field_schema
  File "pydantic/schema.py", line 498, in pydantic.schema.field_type_schema
  File "pydantic/schema.py", line 848, in pydantic.schema.field_singleton_schema
  File "pydantic/schema.py", line 736, in pydantic.schema.field_singleton_sub_fields_schema
  File "pydantic/schema.py", line 563, in pydantic.schema.field_type_schema
  File "pydantic/schema.py", line 947, in pydantic.schema.field_singleton_schema
ValueError: Value not declarable with JSON Schema, field: name='_rule' type=Rule required=True
```

Whereas with Pydantic 1.8 the output is as expected, i.e.
```
{'title': 'RuleData', 'type': 'object', 'properties': {'rule': {'title': 'Rule', 'type': 'array', 'items': {'$ref': '#/definitions/Rule'}}}, 'required': ['rule'], 'definitions': {'Rule': {'title': 'Rule', 'type': 'object', 'properties': {'name': {'title': 'Name', 'type': 'string'}}, 'required': ['name']}}}
```

**Comment by PrettyWood** [comment]
Hi @pgjones 
Bug introduced by https://github.com/samuelcolvin/pydantic/pull/2364. Fix welcome

**Comment by himbeles** [comment]
@PrettyWood, I have contributed a fix here:  #3819

**Comment by himbeles** [comment]
Fixes also #3848

**Comment by lucasfrancaid** [comment]
When the Pull Request of this Issue will be merged?
